### ISSUE TITLE
Kafka sink with ENVELOPE UPSERT cannot handle multiple updates at the same timestamp

### ISSUE THREAD

**Issue by aljoscha** [MaterializeInc/materialize#6526]
When using the RT timestamper and ingesting a lot of data at once it can happen that messages with the same key get assigned the same timestamps. The sink currently panics on those: https://github.com/MaterializeInc/materialize/blob/3dc3354a649993ae739626dfa9b84a754b884462/src/interchange/src/envelopes.rs#L145

For testing, it's even enough to just have two messages in the input topic that have the same key. Also, the key here refers to the `KEY` that is configured on the sink.

I believe the reason for this is that the `combine_at_timestamp()` consumes raw change batches:
https://github.com/MaterializeInc/materialize/blob/9218c98f8577c705db86d410c2e3b2244a425fb1/src/dataflow/src/render/sinks.rs#L97

These don't go through consolidation/arrangement, I think.

Source/sink definitions, for reference, I find it easiest to test things using testdrive:
```
> CREATE SOURCE source_insertavrotest
  FROM KAFKA BROKER '${testdrive.kafka-addr}'
  TOPIC 'upsertavrotest-big'
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  ENVELOPE NONE

> CREATE SINK kafka_sink FROM source_insertavrotest
  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'output-sink-${testdrive.seed}'
  KEY (a)
  WITH (consistency=true)
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  ENVELOPE UPSERT
```

Input are simple messages with schema `(a BIGINT, b BIGINT)`.

**Comment by aljoscha** [comment]
cc @frankmcsherry @umanwizard

**Comment by frankmcsherry** [comment]
A thing I don't know, and we might want to bolt down, is what the intended definition of the upsert sink should be. For other operators that's a thing that I try to do; maybe it makes sense here may not. For example, one worry is that if the intent is "for each key, the value of the most recent update with a positive increment", this is not something that is stable under time compaction. So, e.g. if we restart with an `as_of` used to fast-forward timestamps, we may reach incorrect conclusions about what we should produce (ties were previously broken by time, but that distinction is potentially gone now).

The intent for "collections that have `key` as a unique key" is probably a lot less ambiguous, as at each (key, time) there should be only one value, and it should be clearly presented (and even the retraction). This holds even if we fast-forward times.

**Comment by frankmcsherry** [comment]
Other weirdness examples. If we want to have an upsert sink that reports the most recent value, you would probably expect
```
((frank, val1), time1, +1)
((frank, val2), time2, +1)
```
to result in `(frank, val2)` in the output (assuming `time2 > time1`). What happens if we next receive
```
((frank, val2), time3, -1)
```
That looks like it is just dropping something without replacing it. Should that be a `(frank, None)` in the output to signal dropping the key? That would be weird, because a `SELECT *` from that stream of data would show that a record exists.

Also, what if instead we saw a 
```
((frank, val1), time3, -1)
```
This is a retraction, but not a retraction of the previously emitted value. That probably shouldn't result in a `(frank, None)` output, because `(frank, val2)` is still present in the collection.

Anyhow, strong feels that someone needs to spec out what `ENVELOPE UPSERT` means in contexts where the key is not a unique key, and potentially disallow it / strongly warn if we don't think we have clear behavior there.

**Comment by frankmcsherry** [comment]
Further thoughts: I think there are a sane semantics if the collection is 1. keyed by the upsert key, or 2. append-only. That second case could be made to work too (though, maybe a different implementation? Maybe just "the most recent insert" still works?).

**Comment by aljoscha** [comment]
Two interesting finds: 1) we have zero test coverage for UPSERT sinks, 2) surprisingly, we seem to be handling your example just fine.

I fixed `testdrive` to allow checking for null values, after which this works:
```
> CREATE SOURCE input
  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-input-${testdrive.seed}'
  FORMAT AVRO USING SCHEMA '${schema}' ENVELOPE DEBEZIUM

> CREATE VIEW my_max AS SELECT a, max(b) as b from input GROUP BY a

> CREATE SINK output FROM my_max
  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'output-sink-${testdrive.seed}'
  KEY (a)
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE UPSERT

$ kafka-ingest format=avro topic=input schema=${schema} timestamp=1
{"before": null, "after": {"row": {"a": 1, "b": 1}}}

$ kafka-verify format=avro sink=materialize.public.output
{"a": 1} {"a": 1, "b": 1}

$ kafka-ingest format=avro topic=input schema=${schema} timestamp=1
{"before": null, "after": {"row": {"a": 1, "b": 2}}}

$ kafka-verify format=avro sink=materialize.public.output
{"a": 1} {"a": 1, "b": 2}

$ kafka-ingest format=avro topic=input schema=${schema} timestamp=1
{"before": {"row": {"a": 1, "b": 2}}, "after": null}

$ kafka-verify format=avro sink=materialize.public.output
{"a": 1} {"a": 1, "b": 1}

$ kafka-ingest format=avro topic=input schema=${schema} timestamp=1
{"before": {"row": {"a": 1, "b": 1}}, "after": null}

# Note that a: 1 is the key here while the "nothing" is the null value
$ kafka-verify format=avro sink=materialize.public.output
{"a": 1}
```

**Comment by aljoscha** [comment]
As a first step, I would add a check in `plan_create_sink()` that compares the (user-specified) upsert key against the collection key(s) and bails if the upsert key is not good. I would also add testdrive tests that pin the current behaviour (if we like it). 

Regarding append-only collections: is there a way to know that a collection (relation) is append only. The sink only sees a `RelationDesc`, which contains column types, names, and unique keys.

**Comment by frankmcsherry** [comment]
It is an analysis that we do (`monotonic` is the keyword to search for) but it is probably hard to find. We only currently use it to inform `Reduce` implementation, and so we just stash the results of the analysis in that expression. So. For the moment, probably best not to be blocked on that.

**Comment by aljoscha** [comment]
> Maybe just "the most recent insert" still works?).

So it would be "whatever arrives latest at the sink"? That can be tricky with parallelism.

And if we take the timestamp to decide what is latest we're back at the original problem of multiple updates at the same timestamp.

I would finish https://github.com/MaterializeInc/materialize/pull/6574 (because it also touches the UPSERT code paths) and then put the key checks + tests on top in a follow-up PR.

And then I'll think about what of this we want to document and how.

**Comment by frankmcsherry** [comment]
> So it would be "whatever arrives latest at the sink"? That can be tricky with parallelism.

I think we probably still want "hash partitioned by key" as the upsert output format is offset-ordered, and if one key went to two workers we wouldn't be able to guarantee the right one is written second.

> And if we take the timestamp to decide what is latest we're back at the original problem of multiple updates at the same timestamp.

Right, be we could maybe make an arbitrary decision and have it work? I don't know. I'm not in love with the append-only stuff, but I wanted to put it out there as my complaints about retractions don't apply. I think it's still fair to complain that the fast-forwarded data result in ambiguous outputs if time-ordering gets hidden by fast-forwarding of time.

I'm totally ok with "only allow upsert when keys line up" myself. But probably best to socialize that because it was possibly done more generally for a reason.
